File: Packages/WhoopStore/Sources/WhoopStore/BackupProvenance.swift
import Foundation

/// Build provenance for exports (#1410): a `manifest.json` entry recording which build produced a
/// `.noopbak`, and an `APP_VERSION_CHANGED` event marking each in-app version transition so a SINGLE
/// export answers "what ran when" across the whole retention window. Pure JSON only — the ZIP container
/// and the event write live in the app layer; these are the byte-parity contract with Android.
///
/// Neither is telemetry: both are LOCAL metadata inside the user's own backup, nothing leaves the device.

/// The `manifest.json` entry inside a `.noopbak` — "which build produced this file" (#1410 tier 1).
/// Sibling to `BackupSettings`; not restored, read only to classify a file (helps #746 route by what a
/// file SAYS it is rather than probing tables).
public enum BackupManifest {
    /// Canonical entry name inside the `.noopbak` ZIP. Matches the Android exporter byte-for-byte.
    public static let entryName = "manifest.json"

    /// Deterministic (`.sortedKeys`) JSON: `appBuild`, `appVersion`, `exportedAt` (unix ms), `platform`
    /// ("apple"/"android"), `schemaVersion` (the DB's native schema integer for that platform). The twin
    /// of Android `BackupManifest.json`; keys + kinds identical, values platform-specific.
    public static func json(appVersion: String, appBuild: String, platform: String,
                            schemaVersion: Int, exportedAtMs: Int64) -> String {
        let obj: [String: Any] = [
            "appVersion": appVersion,
            "appBuild": appBuild,
            "platform": platform,
            "schemaVersion": schemaVersion,
            "exportedAt": exportedAtMs,
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: obj, options: [.sortedKeys]) else { return "{}" }
        return String(decoding: data, as: UTF8.self)
    }
}

/// The `APP_VERSION_CHANGED` event appended on an in-app version transition (#1410 tier 2). The one piece
/// no other means can reconstruct: a single export then answers "what ran when" retroactively.
public enum AppVersionEvent {
    /// The `event.kind` for a version transition. Rides the existing `event` table (`payloadJSON` string).
    public static let kind = "APP_VERSION_CHANGED"

    /// Record a transition only when a PRIOR version was seen AND it differs — a first launch (nil prior)
    /// records nothing (there is no transition), so the table never carries a spurious "from null" row.
    public static func shouldRecord(lastSeen: String?, current: String) -> Bool {
        guard let lastSeen, !lastSeen.isEmpty else { return false }
        return lastSeen != current
    }

    /// Deterministic (`.sortedKeys`) payload: `from`, `schemaVersion`, `to`. Twin of Android
    /// `AppVersionEvent.payloadJson`.
    public static func payloadJson(from: String, to: String, schemaVersion: Int) -> String {
        let obj: [String: Any] = ["from": from, "to": to, "schemaVersion": schemaVersion]
        guard let data = try? JSONSerialization.data(withJSONObject: obj, options: [.sortedKeys]) else { return "{}" }
        return String(decoding: data, as: UTF8.self)
    }
}

File: android/app/src/main/java/com/noop/data/BackupProvenance.kt
package com.noop.data

/**
 * Build provenance for exports (#1410): a `manifest.json` entry recording which build produced a
 * `.noopbak`, and an `APP_VERSION_CHANGED` event marking each in-app version transition so a SINGLE
 * export answers "what ran when" across the whole retention window. Pure JSON only — the ZIP container
 * and the event write live in the app layer; this is the byte-parity twin of Swift `BackupProvenance.swift`.
 *
 * Neither is telemetry: both are LOCAL metadata inside the user's own backup, nothing leaves the device.
 *
 * The JSON is a minimal, sorted-key object built by hand so it is byte-identical to Swift's
 * `JSONSerialization(.sortedKeys)` and testable in a plain JVM (no `org.json` stub).
 */
private fun jsonString(value: String): String = buildString(value.length + 2) {
    append('"')
    value.forEach { char ->
        when (char) {
            '"' -> append("\\\"")
            '\\' -> append("\\\\")
            '\b' -> append("\\b")
            '\u000c' -> append("\\f")
            '\n' -> append("\\n")
            '\r' -> append("\\r")
            '\t' -> append("\\t")
            else -> if (char.code in 0x00..0x1f) {
                append("\\u")
                append(char.code.toString(radix = 16).padStart(length = 4, padChar = '0'))
            } else {
                append(char)
            }
        }
    }
    append('"')
}

private fun sortedJsonObject(entries: List<Pair<String, Any>>): String =
    entries.sortedBy { it.first }.joinToString(separator = ",", prefix = "{", postfix = "}") { (k, v) ->
        val value = when (v) {
            is String -> jsonString(v)
            else -> v.toString()   // Int/Long → unquoted numeric literal
        }
        "\"$k\":$value"
    }

/** The `manifest.json` entry inside a `.noopbak` — "which build produced this file" (#1410 tier 1). */
object BackupManifest {
    /** Canonical entry name inside the `.noopbak` ZIP. Matches the Swift exporter byte-for-byte. */
    const val ENTRY_NAME = "manifest.json"

    /** Twin of Swift `BackupManifest.json`; keys + kinds identical, values platform-specific. */
    fun json(appVersion: String, appBuild: String, platform: String, schemaVersion: Int, exportedAtMs: Long): String =
        sortedJsonObject(
            listOf(
                "appVersion" to appVersion,
                "appBuild" to appBuild,
                "platform" to platform,
                "schemaVersion" to schemaVersion,
                "exportedAt" to exportedAtMs,
            )
        )
}

/** The `APP_VERSION_CHANGED` event appended on an in-app version transition (#1410 tier 2). */
object AppVersionEvent {
    /** The `event.kind` for a version transition. Rides the existing `event` table (`payloadJSON` string). */
    const val KIND = "APP_VERSION_CHANGED"

    /** Record only when a PRIOR version was seen AND it differs — first launch (null/blank prior) records nothing. */
    fun shouldRecord(lastSeen: String?, current: String): Boolean =
        !lastSeen.isNullOrEmpty() && lastSeen != current

    /** Twin of Swift `AppVersionEvent.payloadJson`: deterministic `from`, `schemaVersion`, `to`. */
    fun payloadJson(from: String, to: String, schemaVersion: Int): String =
        sortedJsonObject(listOf("from" to from, "to" to to, "schemaVersion" to schemaVersion))
}
